docs/negation-types.md

```diff
@@ -0,0 +1,86 @@
+
+A [cofinite set](https://en.wikipedia.org/wiki/Cofiniteness) arises when you have a negation type in conjunction with intersection types. For example, `string & ~"a"` is a cofinite string set that accepts any `string` excluding `"a"`. This typically happens all the time with type refinements, where `typeof(x) == "string" and x ~= "a"` gives rise to the type `string & ~"a"`.
+
+## Design
+
+### Syntax
+
+The notation is `~T` where `T` is a type annotation. It binds tightly to the type it negates, so `~T | U` is `(~T) | U`, not `~(T | U)`. There's really not much to it, syntax-wise. It has the same precedence as the `not` expression, which is why `~T | U` is `(~T) | U` in the same way `not t or u` is `(not t) or u`.
+
+We propose to allow users to write such a type. Formally:
+
+```ebnf
+ty ::= `~` ty
+     | name [`<` [ty] {`,` ty} `>`]
+     | ty {`|` ty}
+     | ty {`&` ty}
+     | `(` ty `)`
+     | `typeof` `(` exp `)`
+     | table_ty    (* omitted for brevity *)
+     | function_ty (* omitted for brevity *)
+```
+
+### Semantics
+
+The basis set we want to perform set exclusion on is `unknown`, _not_ `any`. This means given the type `~"a"`, it is equivalent to the type `unknown & ~"a"`.
+
+Choosing `unknown` as the basis makes a lot of things fall in place properly, including the property we wish to maintain where `~~T` is equivalent to `T`. It's crucial that we handle negation of error types properly, otherwise double negations won't actually be consistent. It'd allow error suppression to be improperly suppressed! See below, where `~~any` ends up as `unknown` instead of `any`.
+
+1. `~~any`
+2. `~~(unknown | *error*)`
+3. `~(~unknown & ~*error*)`
+4. `~(never & unknown)`
+5. `~never`
+6. `unknown` (not equivalent to `any`!)
+
+As you can see, since through the series of rewrites `~~any` did not produce a type equivalent to `any`, our basis set must be `unknown`, and negation of an error suppression is still an error suppressing type. Therefore `~*error*` is `*error*`. This fixes the double-negation inconsistency:
+
+1. `~~any`
+2. `~~(unknown | *error*)`
+3. `~(~unknown & ~*error*)`
+4. `~(never & *error*)`
+5. `~never | ~*error*`
+6. `unknown | *error*` (`any` is an alias to `unknown | *error*`, so equivalence is achieved)
+
+### Restrictions
+
+We currently only support negation of _testable_ types. Intuitively, a testable type is one that can be tested by type refinements at runtime. This includes singleton types, primitive types, and classes. Of course, unions and intersections can be negated as well if and only if all of its constituent types are also testable types, due to the distributivity property of unions and intersections.
+
+Since types like `{ p: P }` and `(T) -> U` are structural types and the runtime offers no way to test for them, they cannot be negated. An attempt to negate this will just turn the negation type into an error suppressed type and also a type error. This means `~{ p: P }` and `~(T) -> U` are nonsensical.
```

Comment:
We should probably explicitly error if a user writes an annotation that is the negation of a structural type.

Reply:
There was "and also a type error" but I can extract that into a separate sentence in this paragraph.

Reply:
Word ordering is super important and it's rather annoying. Function type. Type function. Refinement type. Type refinement.